### ISSUE TITLE
Library structure

### ISSUE THREAD

**Issue by alteous** [gltf-rs/gltf#13]
This thread exists as a place to discuss the modules, functions, data structures, and their names included in the library. In version `1.0.0` names must be agreed upon and remain immutable for backward compatibility. Until then, we should experiment as much as possible and find out what works best.

#### gltf

 - The main crate, intended for navigating glTF JSON.
 - Intentionally limited in scope.

#### gltf-json

 - Child crate for `gltf`.
 - Exists primarily to reduce compile times for `gltf`.
 - `semver` matches that of `gltf` for convenience but is otherwise unstable.

 #### gltf-importer

- Reference importer with a simple API.
- May be used by run times to read from the file system.
- Handles base 64 decoding and distinguishing binary glTF on behalf of the user.
- Optionally loads buffers and images (to do).

#### gltf-utils

- Utility functions to complement the `gltf` crate.
- Addresses cases where access to pre-existing buffer / image data is required.
- ~~May remain unstable post `gltf 1.0`.~~

**Comment by bwasty** [comment]
I think the main import function could use a variant that takes a [`std::io::Read`](https://doc.rust-lang.org/nightly/std/io/trait.Read.html) for input and uses [`serde_json::from_reader()`](https://docs.serde.rs/serde_json/fn.from_reader.html) for deserializing. This would
* make reading from a file more efficient (no full copy in memory needed)
* enable loading models directly from the network (the `Response` structs of both `hyper` and `reqwest` implement `Read`)

**Comment by alteous** [comment]
This is not a bad idea, but there is one caveat, namely the dependence on the directory path for buffer and image data sources. If there is a reasonable workaround for this then I'm 100% with you.

Note that:
 - In the `v1` module, generic `Read` is currently when [importing binary glTF](https://github.com/alteous/gltf/blob/incoming/src/v1/import.rs#L39).
 - The specification states under [design goals](https://github.com/KhronosGroup/glTF/tree/2.0/specification/2.0#design-goals) that glTF is not a streaming format.

**Comment by alteous** [comment]
I suppose we could provide a `import_binary()` function that only imports binary or embedded glTF formats.

**Comment by bwasty** [comment]
I understood the section about streaming to be more about starting rendering before everything is loaded rather then whether files are loaded directly from disk.
Especially since (non-embedded) buffers and images are referenced via [relative URIs](https://github.com/KhronosGroup/glTF/tree/2.0/specification/2.0#uris), which doesn't restrict the scheme to `file` as far as I see. Although:
> Clients are free to support other schemes (such as `http://`) depending on expected usage.

suggests that `file` is implicitly the default.

Maybe we could abstract over URIs and have (user defined) implementations for different schemes that handle the actual loading. 
My end goal would be that it's possible to compile to WebAssembly and still load any kind of glTF. This might even mean a callback to JavaScript to use the [Fetch API](https://developer.mozilla.org/en/docs/Web/API/Fetch_API) asynchronously.

Another benefit of an API that supports this would be that even local buffer/image references could be loaded in parallel for maximum performance.

**Comment by alteous** [comment]
> I understood the section about streaming to be more about starting rendering before everything is loaded ...

That would make a lot more sense, I misunderstood the meaning of streaming, 

> Maybe we could abstract over URIs ...

The scope of the crate is growing larger than I originally planned. When you say
abstraction over URIs, do you have anything particular in mind?

My first thoughts were a simple callback with the URI passed as a parameter. The
user would then return a `Vec<u8>`. This approach seems a bit too simplistic to be future-proof.

My second thoughts were some kind of trait, for example: 

```rust
trait Reader {
    type Error;
    fn read_buffer(&mut self, buffer: &Buffer) -> Result<Vec<u8>, Self::Error>;
    fn read_image(&mut self, image: &Image) -> Result<Vec<u8>, Self::Error>;
}
```

The user would then write something like this:

```rust
struct Reader;

impl Reader for Reader {
    type Error = ...;
    fn read_buffer(&mut self, buffer: &Buffer) -> Result<Vec<u8>, Self::Error> {
        if buffer.uri().starts_with("file://") {
            // Read from file system
        } else if buffer.uri().starts_with("http://") {
            // Download from internet
        } else {
            // Unsupported scheme
            Err(...)
        }
    }

    fn read_image(&mut self, image: &Image) -> Result<Vec<u8>, Self::Error> {
        // Similar to above
    }
}

fn main() {
    let gltf = gltf::v2::import("Foo.gltf", Reader);
}
```

This seems like a lot of work on the user's behalf. Perhaps we could provide a
basic synchronous implementation for convenience:

```rust
fn main() {
    let gltf = gltf::v2::import("Foo.gltf", gltf::v2::reader::Simple);
}
```

**Comment by alteous** [comment]
Third thoughts: The above is no good since the original discussion was about `import()` taking generic `Read`. So perhaps the user should write something like this instead:

```rust
/// A glTF source. Calling `Read::*()` means reading the .gltf file.
/// `read_buffer()` and `read_image()` are extra required methods needed to read
/// the data referenced by the .gltf file.
trait Source: Read {
    type Error;
    fn read_buffer(&mut self, buffer: &Buffer) -> Result<Vec<u8>, Self::Error>;
    fn read_image(&mut self, image: &Image) -> Result<Vec<u8>, Self::Error>;
}

enum Reader {
    /// glTF hosted on the file system.
    File(PathBuf),

    /// glTF hosted on the world wide web.
    Http(Url),
}

impl Read for Reader {
    ...
}

impl Source for Reader {
    type Error = ...;
    fn read_buffer(&mut self, buffer: &Buffer) -> Result<Vec<u8>, Self::Error> {
        if buffer.uri().starts_with("file://") {
            // Read from FS
        } else if buffer.uri().starts_with("http://") {
            // Download from WWW
        } else {
            // Unsupported scheme
            Err(...)
        }
    }

    fn read_image(&mut self, image: &image) -> Result<Vec<u8>, Self::Error> {
        // Similar to above
    }
}

fn main() {
    let gltf_from_fs = gltf::v2::import(Reader::File("./dir/Foo.gltf"));
    let gltf_from_www = gltf::v2::import(Reader::Http("example.com/Foo.gltf"));
}
```

**Comment by bwasty** [comment]
That's pretty much what I had in mind, though I hadn't thought it out this far. To avoid the if-else cascade in `read_buffer` (... there's also the "data" URI case), I thought of several structs, `FileReader` and `HttpReader` that would each implement `Source`.

I wonder though, who would actually call `read_buffer`/`read_image`? The `import` function or the user afterwards? If the latter, even asynchronous/parallel loading could be easily done by the user with a different approach, e.g. traversing all buffers/images collecting URIs, and then using something like the new [Async Hyper Client](https://mgattozzi.com/hyper-client).

Also, where would the (down)loaded data be saved? Extra fields on `Buffer`/`Image` or up to the user?

About the compile times: I was already surprised how slow `cargo test` is after making small changes to the lib - takes about 55seconds. Is that related?

**Comment by alteous** [comment]
> who would actually call read_buffer/read_image

I've been planning a 'rustic' wrapper for the crate for a while. You can browse the ideas in this new [`wrapper`](https://github.com/alteous/gltf/tree/wrapper) branch. One of the responsiblities of the wrapper would be to load buffers ([example](https://github.com/alteous/gltf/blob/wrapper/src/v2/root.rs#L197)) and images on behalf of the user.

>  The import function or the user afterwards?

I'd say the `import` function as the name suggests. The `Reader` trait would aliow the user to customise data acquisition with the async hyper client or otherwise.

> where would the (down)loaded data be saved

This has been a pain point in the wrapper idea because the Rust ownership rules start to get in the way. The current way the wrapper handles this is to preload everything upon creation. This seems like a bad idea though because of blocking I/O, denial of service, et cetera. External data should ideally be lazily loaded but as far as I'm aware this means the user would have to make a place to store the data and pass `&mut` around, thus removing any opportunities for parallelism.

> About the compile times: I was already surprised how slow cargo test is after making small changes to the lib - takes about 55seconds. Is that related?

Maybe. I'd suggest using `cargo check` if you're just making small changes. I find it difficult to reason about the performance of `rustc` at the moment.

**Comment by bwasty** [comment]
> I'd say the `import` function as the name suggests. The `Reader` trait would aliow the user to customise data acquisition with the async hyper client or otherwise.

Sounds like a good default, but I think the `Reader` trait won't  properly work for async loading since it synchronizes the reading operations. The client may load asynchronously internally, but for parallel loading you'd need a different interface I think (I looked a bit into using `futures` directly, or `rayon`).

>  ... the Rust ownership rules start to get in the way ... External data should ideally be lazily loaded but as far as I'm aware this means the user would have to make a place to store the data and pass &mut around, thus removing any opportunities for parallelism.

Might `RefCell` or `Arc` help? I don't have any experience with the advanced pointer types, but this reminds me a bit of what I read [here](https://mgattozzi.com/refactor-rust) (paragraph in the middle). 

On the other hand, perhaps it would be better to leave some of the details to the user and just provide basic helper functions and structs. For example, lazy loading could mean different things:
* load data for individual nodes when they should be rendered (or first accessed)
* load everything related to the current scene
* do frustum culling and load everything visible (I know, sounds like streaming, but the spec also says the `binary data in glTF is inherently streamable` and the min/max properties on accessors should make this easy)

> I'd suggest using `cargo check` if you're just making small changes.

Even better is the [RLS](https://github.com/rust-lang-nursery/rls) (though it was broken in nightly unfortunately when I was working on gltf...)

**Comment by alteous** [comment]
> Reader trait won't properly work for async loading since it synchronizes the reading operations

This is a very good point.  If an `import_async` function was provided I'm not sure how the data would be lazily-loaded, even with futures - I'll have a think about it. For the `1.0.0` release it would be worth just concentrating on the synchronous case.

> Might RefCell or Arc help? 

Let's not go there if possible. It would solve some problems but also create new problems, for example making sure there is only one writer at runtime.

> Even better is the RLS ...

Thanks, I'll check that out.

**Comment by alteous** [comment]
Some more points:

> load data for individual nodes when they should be rendered (or first accessed)

There are likely to be many nodes, and if they're all reading little segments from say the file system then I doubt that would perform well, but that's only speculation.

> load everything related to the current scene

There is a further point to be raised here: Since glTF can contain multiple scenes then perhaps reading the .gltf / .glb file and loading the referenced data should be two separate operations. For example:

 ```rust
// Calls `serde_json::from_whatever` and performs some validation
fn import(path: Path) -> Gltf { ... }

impl Gltf /* or `Scene` in the wrapper */ {
     // Loads the scene data
     //
     // N.B. taking 'out' parameters is not recommended by the Rust API guidelines
     // but `std::io::Read` does it for reasons similar to one's were facing
     fn load_scene(dest: &mut SceneData) -> LoadResult<()> { ... }
}
 ```

> On the other hand, perhaps it would be better to leave some of the details to the user and just provide basic helper functions and structs. 

I used to have this mindset but then I began to think if the user has this burden then they are unlikely (at least in the short term) to adopt glTF into their project. The wrapper exists alongside the 'raw' data structures to allow any user to do their own thing if they *really* want to.

> do frustum culling and load everything visible (I know, sounds like streaming, but the spec also says the binary data in glTF is inherently streamable and the min/max properties on accessors should make this easy)

That sounds like a optimisation that's out of scope for the `1.0.0` release however it may be an interesting long-term goal.

**Comment by bwasty** [comment]
> There are likely to be many nodes, and if they're all reading little segments from say the file system then I doubt that would perform well, but that's only speculation.

I'd expect that if there's only one .bin file as in all the examples I've seen so far, only the first node accessed would read from the file system, later ones would get the data from memory.

Though I also like the separate scene loading interface.

**Comment by alteous** [comment]
Here's a [draft of the new import strategy](https://github.com/alteous/gltf/blob/wrapper/src/import.rs). I'm still not sure how to handle async yet.

**Comment by bwasty** [comment]
Looks/works good so far. Two comments:
* is there a `meshes` function missing on `Gltf`? I tried ```let mesh = &gltf.meshes().nth(0);```, since that worked with `buffers`.
* The .glb format is not an extension anymore, so the error type should no be `ExtensionUnsupported`

**Comment by alteous** [comment]
Both the mesh iterator on `Gltf` and the .glb format is unintentionally missing. I'll fix this very soon.

I'm pleased to hear the wrapper is working well.

**Comment by bwasty** [comment]
Finally got to a [first successful render](https://user-images.githubusercontent.com/1647415/27515941-44db32ca-59af-11e7-8b04-9bfe30aa62ef.png) of `Box.gltf` in my viewer (took a while since started [this](https://github.com/bwasty/learn-opengl-rs/) in the mean time to get toward a PBR renderer). So here are some more comments:
* The `position/normal/tex_coord` iterators on gltf::mesh::Primitive are really convenient, much nicer/shorter than my previous attempt
  - nitpick: I was a bit confused for a moment that they're singular. It makes sense when it's called "position attribute", but without "attribute" in the name I'd find it much more intuitive in the plural.
* I've run into some missing `Debug` implementations on some things I tried:
```
println!("pos: {:?}", primitive.position());
```
```
assert_eq!(primitive.mode(), Mode::Triangles);
```

I'll try writing a network-based `Source` as soon as I have some basic geometry/texture handling done.

**Comment by bwasty** [comment]
Another thing: loading `minimal.gltf` fails with 
`Error: Io(Error { repr: Os { code: 2, message: "No such file or directory" } })`
I guess it's because data URIs are not supported yet for buffers (removing it causes JSON validation errors), but it would be nice if the error contained the file name.

Perhaps we could track somewhere what already works/what is missing?

**Comment by alteous** [comment]
> Finally got to a first successful render of Box.gltf in my viewer

That's wonderful news! It's great to see the crate in action.

> The position/normal/tex_coord iterators on gltf::mesh::Primitive are really convenient, much nicer/shorter than my previous attempt

They are pretty nifty if I may say so myself, and especially so if used with [`multizip`](https://crates.io/crates/multizip). 

> nitpick: I was a bit confused for a moment that they're singular. It makes sense when it's called "position attribute", but without "attribute" in the name I'd find it much more intuitive in the plural.

The only reason why it's named this way is because then it matches the attribute name in the specification. I completely agree that they're confusing, so let's change them to plurals.

> I guess it's because data URIs are not supported yet for buffers

It's because the URI contains embedded base64 data, which is at the top of my 'to do' list to implement.

> Perhaps we could track somewhere what already works/what is missing?

That's a good idea. I think it would be best to do this using GitHub issues with custom tags. Many Rust projects seem to do this successfully.

> I'll try writing a network-based Source as soon as I have some basic geometry/texture handling done.

If you have the time, feedback would be appreciated.

**Comment by bwasty** [comment]
> ... especially so if used with multizip.

Interesting alternative, I wanted to use [`itertools::izip!`](https://docs.rs/itertools/0.6.0/itertools/macro.izip.html)

**Comment by alteous** [comment]
[It's almost the same](https://docs.rs/itertools/0.6.0/src/itertools/lib.rs.html#181).

**Comment by alteous** [comment]
By the way there's a new version of the wrapper interface. [GLB is supported now](https://github.com/alteous/gltf/blob/wrapper/src/import/binary.rs) - the implementation was much more difficult to write than I'd expected.

**Comment by bwasty** [comment]
>Perhaps we could track somewhere what already works/what is missing?
>>That's a good idea. I think it would be best to do this using GitHub issues with custom tags. Many Rust projects seem to do this successfully.

My first Idea was a single issue with a few checkboxes.

> GLB is supported now

Nice!

**Comment by alteous** [comment]
> My first Idea was a single issue with a few checkboxes.

That would be much simpler. I'll do that for now.

**Comment by alteous** [comment]
> My first Idea was a single issue with a few checkboxes.

How about an issue with a list of all the things that are in a working state? The rest can have individual issues so that we can discuss the details easily.

**Comment by bwasty** [comment]
Found another surprising thing: I wanted to return a `Gltf` from a loading function, but couldn't because the liftetime was bound to the `Importer` instance (which was local to the function). Looking at the code I understand why, but I'd naively expect that `Gltf` owns all the data and can be moved.

**Comment by alteous** [comment]
This is a good point. Until now I've believed that simply returning `Gltf<'static>` would suffice but this is clearly not the case. If `Gltf` owned its buffer and image data instead of borrowing it from an `Importer` then my original supposition should hold, I.e. you should be able to return `Gltf<'static>` without the borrow checker grumbling. I'll make an update on this tomorrow.

**Comment by alteous** [comment]
**Bad news**

I was wrong about returning `Gltf<'static>` from `Importer`. The reason is illustrated below:

```rust
impl Owner {
    fn borrow<'a>(&'a self) -> Borrowed<'a> {
        // self is borrowed even if 'a is the 'static lifetime
    }
}
```

**Good news**

This problem was easily solved by instead having two flavours of the importer - I called them `Static` and `ZeroCopy`. The changes are in commit https://github.com/alteous/gltf/commit/ef828be068b9a31e1bf52d9832af2e637ae4361b.

**Comment by alteous** [comment]
Also, `Gltf` now owns its buffer and image data - see https://github.com/alteous/gltf/commit/2362574250bd8c05f67ff985676c0c65b5c3adc9 for details.

**Comment by bwasty** [comment]
With the increasing number of possible options, it might be worth considering the [builder pattern](https://github.com/brson/rust-api-guidelines#c-builder). Something along the lines of:
```
let importer = ImporterBuilder::new()
   .source(NetworkSource::new("http://example.com/foo.gltf")
   .load_binary_data(true), // with `false`, only load the json, leave the rest to the user
   .mode(Mode::Async) // or ZeroCopy
   .lazy_loading(Lazy::Off) // or Scene/All
   .validation(Validation::Complete) // or Minimal
   .generate_missing(Semantic::Normals) // can be repeated with Semantic::Tangents
   .build();
```
This might also make it easier to add new options later without breaking the API.

**Comment by alteous** [comment]
Could [`mint`](https://github.com/kvark/mint) be useful for returning math types?

**Comment by alteous** [comment]
Did you try `node.matrix().into()`?

**Comment by alteous** [comment]
We could make functions more generic, maybe?

```rust
impl Node {
    fn matrix<T>(&self) -> T
        where T: From<[f32; 4]>
    {
        self.json.matrix.into()
    }
}

**Comment by bwasty** [comment]
> Could mint be useful for returning math types?

It could be if it was actually integrated with the libraries - https://github.com/kvark/mint/issues/11 suggests that it's not very successfull so far. I also have the feeling that most people now use `cgmath` for CG.

> Did you try node.matrix().into()?

Yes, same message as for `Matrix4::from`: ```the trait `std::convert::From<[f32; 16]>` is not implemented for `cgmath::Matrix4<f32>```. According to the docs it should work as far as I understand, but it doesn't somehow...

**Comment by alteous** [comment]
It seems there is an `impl<'a, S> From<&'a [S; 16]> for &'a Matrix4<S>` but not an `impl<S> From<[S; 16]> for Matrix4<S>` which is bizarre.

**Comment by alteous** [comment]
The crate received some valuable criticism recently on the [#rust-gamedev](https://botbot.me/mozilla/rust-gamedev/2017-08-04/?msg=89404927&page=3) IRC channel. Below is a summary of the comments received.

1. @tomaka criticised the use of `futures` due to compatibility and dependency issues when versions mismatch. This is a known issue but not one the crate has ever fully acknowledged.
2. @tomaka and @Ralith criticised the fact that the crate performs I/O as well as parsing and providing abstractions over data. I don't see this as a huge problem, as long as it's optional and there is a reasonable alternative available.
3. @Ralith heavily criticised the fact that `gltf` performs image decoding, arguing that it should not the job of `gltf` to perform this task at all. I am inclined to agree since it has been a source (heh) of many problems with minimal benefit.
4. @Ralith also briefly mentioned the problem of 'how to deal with extensions'. The current approach is to support official extensions as and when they are required, relying on pull requests to implement the relevant functionality. This a simple but perhaps inflexible approach.
5. My impression from the discussion was that potential users are, first and foremost, interested in a glTF parser, and are irritated by the way the crate is trying to do 'too much' on their behalf. Although it is possible to use `gltf` simply as a JSON parser this is currently far from obvious.

To address these criticisms, I propose the following.

- Split the crate into smaller sub-crates that perform specialised tasks, for example, JSON parsing, glTF tree iteration, and buffer / image data importing. This would incur:
   - Moving the `json` module into its own `gltf-json` crate.
   - Moving the `import` module into its own `gltf-importer` crate.
- Change the strategy `Gltf` (the wrapper type) receives its buffer data.
- Remove `image_data` from `Gltf` entirely.
- Outside of `gltf-importer`, leave I/O completely up to the user.

I encourage further discussion.

**Comment by vitvakatu** [comment]
@alteous 
I'm outsider for this project, but I interested in it, as `gltf` is going to be a de facto standard format for modern 3d applications.
I like your proposal about splitting the crate into several sub-projects, I hope it would be hard.

Btw, `mint` has been integrated into `cgmath` successfully. I can take care of creating a PR for `gltf` if you don't mind.

**Comment by alteous** [comment]
@vitvakatu Thanks for your interest!

> Btw, mint has been integrated into cgmath successfully. I can take care of creating a PR for gltf if you don't mind.

Of course. In case I've misunderstood, would the intention be to replace vector types with `mint` types, i.e. `[f32; 3] -> Vector3<f32>` etc?

**Comment by vitvakatu** [comment]
@alteous only public APIs will be changed, yes, `[f32; 3]` would be replaced with `mint::Vector3<f32>` or `mint::Point3<f32>`.
I made an issue for discussing and tracking progress #67

**Comment by Ralith** [comment]
> I also have the feeling that most people now use cgmath for CG.

I can't speak for everyone, but I am very happy with nalgebra.

> Change the strategy Gltf (the wrapper type) receives its buffer data.

I'm not deeply familiar with the current API (or glTF itself for that matter) but (as alluded to in IRC) it's not at all obvious to me that the root `Gltf` type needs to have access to buffer data at all, any moreso than it needs access to image data. Anywhere the JSON parser currently yields a struct containing a `&Gltf` and an index referencing to some binary data contained within it, the parser could instead yield the index alone and let the user worry about constructing and filling in a table of buffers/views/accessors/images/meshes/etc in whatever manner they like, if indeed they want to at all. So long as the parser provides access to the information necessary to load indirectly referenced data, this should be straightforward to use. You could even provide a helper that just loads everything up front into a simple struct for simple usecases and demos.

If you want to continue to provide friendly accessors like `gltf::mesh::Mesh`, you'd just replace `fn new(gltf: &'a Gltf, index: usize, json: &'a Mesh) -> Self` with something like `fn new(accessor: &'a Accessor<'a>, json: &'a Mesh) -> Self` where `Accessor` is constructed from a `&'a [u8]` (taking the place of a buffer view) and a `gltf::json::accessor::Accessor`.

**Comment by bwasty** [comment]
About splitting into sub-crates: `gltf-json` and `gltf-import` sound reasonable. But we should try not to create too many sub-crates and make sure each is useful on its own. For example, `gltf-derive` doesn't seem useful on it's own and ideally shouldn't show up on crates.io.

Apart from that, I agree with the general direction. I personally like the convenience the crate provides beyond just parsing, but I have no problem moving some of that to my viewer/engine.

**Comment by alteous** [comment]
> I personally like the convenience the crate provides beyond just parsing

As do I. I'd like to redesign the way the wrapper library receives data so that we still have these conveniences. One idea I had this morning was to introduce a `Load` trait that operates on a `Source`. The user could then choose how much data is prepared. The 

```rust
trait Source {
    /// Customisable error type - reference implementation provided by `gltf-importer`.
    type Error;

    /// Returns `Ok(())` if the data is prepared. The data should be loaded by the `Source`
    /// implementation if unavailable, returning `Err(...)` if the data cannot be sourced.
    ///
    /// The user is expected to persist the buffer for the lifetime of the loaded data.
    /// Hopefully the borrow checker can guarantee this for us.
    fn prepare(&mut self, buffer: usize) -> Result<(), Self::Error>;

    /// For the wrapper library to retrieve buffer data without taking `&mut`.
    fn retrieve<'a>(&'a self, buffer: usize) -> &'a [u8];
}

/// Internal trait for converting `Unloaded<T>` to `Loaded<T>`.
trait Load {
    /// The loaded result type.
    type Loaded;
    /// Recurses down the hierarchy, calling `prepare()` on every item that will need to be
    /// ready for the wrapper.
    fn load<S: Source>(self, source: &mut S) -> Result<Self::Loaded, S::Error>;
}

/// Wrapper type for items that may not have their data ready.
struct Unloaded<T: Load> {
    item: T,
}

impl<T: Load> Load for Unloaded<T> {
    type Loaded = T::Loaded;
    fn load<S: Source>(self, source: &mut S) -> Result<Self::Loaded, S::Error>;
}
```

**Comment by alteous** [comment]
Now the advantage of this idea is that you can call `load()` on, say, a `Node`, and it will prepare all the data necessary for that `Node` and all of its children. But, if the user wants more fine-grained control over what's loaded, they can call `load()` on smaller items like `Mesh` or `Skin`. The `Unloaded` version of a struct doesn't have convenience features like attribute iteration.

**Comment by alteous** [comment]
> gltf-derive doesn't seem useful on it's own and ideally shouldn't show up on crates.io

Unfortunately it has to be in its own crate because it's a `proc_macro`. Furthermore, I tried packaging it up with `gltf` but crates.io did not like that at all. It's not a huge issue and [lots of other crates do this](https://crates.io/search?q=derive) - scan for 'internal' crates.

**Comment by alteous** [comment]
Actually, I think [this idea](https://github.com/alteous/gltf/issues/13#issuecomment-320434394) is overly complex. A much simpler approach might be to allow the user to 'query' what data is needed on individual objects, let the user source the data themselves, and provide the data back to the object. This pseudo-code should briefly explain the idea:

```rust
/// A source of buffer data.
trait Source {
    /// N.B. this *must not* fail - the user is expected to
    /// ensure that all the data is available.
    fn source_buffer<'a>(&'a self, buffer: Buffer<'a>) -> &[u8];
}

/// New: Wrapper type for a glTF object with loaded data.
struct Loaded<'a, T> {
    /// The wrapper item - `Mesh` in this example.
    item: T,

    /// User data that (hopefully) contains all the data needed by the item
    /// and all of its children.
    source: &'a Source,
};

/// Same as before.
struct Mesh<'a> {
    gltf: &'a Gltf,
    json: &'a json::Mesh,
    index: usize,
}

impl Mesh {
    /// Returns an `Iterator` over the mesh primtives, none
    /// of which have any data.
    fn primitives(&self) -> Primitives<'a> {
        ...
    }

    /// Returns a `Vec` of all the buffers required by the
    /// user to be loaded.
    fn query(&self) ->Vec<Buffer> {
        vec![Buffer(3), Buffer(1), Buffer(4)]
    }
}

impl Loaded<Mesh> {
    /// Returns an `Iterator` over the mesh primitives, *all* of
    /// which have their data, provided the user initialised the
    /// buffer source correctly.
    fn primitives(&self) ->LoadedPrimitives<'a> {

    }
}

struct MySource {
    /// For example.
    buffers: Vec<Vec<u8>>,
}

impl Source for MySource {
    fn source_buffer<'a>(&'a self, buffer: Buffer<'a>) -> &[u8] {
        &self.buffers[buffer.index()]
    }
}

enum Gltf {
    Standard(Gltf),
    Binary(Gltf, Vec<(u32, Vec<u8>)>),
}

fn main() {
    let mut my_source = MySource;
    let gltf = match gltf::parse(include_bytes!("example.glb")) {
        Gltf::Standard(gltf) => gltf,
        Gltf::Binary(gltf, embedded_buffers) => {
            my_source.store(embedded_buffers);
            gltf
        },
    };
    let unloaded_mesh = gltf.meshes().nth(0).unwrap();
    my_source.load_buffers(unloaded_mesh.query());
    let mesh = unloaded_mesh.load(&my_source);
    for loaded_primitive in mesh.primitives() {
        let indices = loaded_primitive.indices_u32().unwrap().collect();
        // you get the picture...
    }
}
```

**Comment by alteous** [comment]
Also, we can make `query` a trait method and use the same trick as `Validate` to recurse through the heirarchy.

**Comment by Ralith** [comment]
That looks similar to what I had in mind. However, I'd still much rather do away with the `Source` trait entirely and have the parser just yield indexes for everything, and let the user construct `Loaded<T>` from a specific buffer at the last possible moment--no implicit assumptions about entire subtrees being fully available. For example, I wouldn't want to be required to load buffers used only for extensions I don't implement, or data which is irrelevant to a processing task I'm implementing

**Comment by alteous** [comment]
Good news.
>let unloaded_mesh = gltf.meshes().nth(0).unwrap();
    source.load_buffers(unloaded_mesh.query());
    let mesh = unloaded_mesh.load(&source);

Good news: [this can be done in one line!](https://play.rust-lang.org/?gist=a67a2d02966b1e6669fa2652a8921d9a&version=stable).

Contents of example:
```rust
struct Loader {
    data: Vec<u8>,
}

#[derive(Debug)]
struct LoadError;

trait Source {
    fn source(&self, buffer: usize) -> &[u8];
}

trait Query {
    fn query(&self) -> Vec<usize>;
}

struct Item {
    required_buffer: usize,
}

impl Query for Item {
    fn query(&self) -> Vec<usize> {
        vec![1, 2, 3]
    }
}

struct Loaded<'a, T> {
    item: T,
    source: &'a Source,
}

impl<'a> Loaded<'a, Item> {
    fn print(&self) {
        println!("{:?}", self.source.source(self.item.required_buffer));
    }
}

impl Loader {
    fn load<T: Query>(&mut self, item: T) -> Result<Loaded<T>, LoadError> {
        for buffer in item.query() {
            self.data.push(buffer as u8);
        }
        Ok(Loaded {
            item,
            source: self,
        })
    }
}

impl Source for Loader {
    fn source(&self, _buffer: usize) -> &[u8] {
        self.data.as_slice()
    }
}

fn main() {
    let mut loader = Loader { data: vec![] };
    let unloaded_item = Item { required_buffer: 123 };
    let item = loader.load(unloaded_item).expect("loading error");
    item.print();
}
```

**Comment by alteous** [comment]
The above decouples buffer loading from tree traversal. The `Loader` type is nothing special - a similar implementation could be written the user if specialisation is required. The crate could provide a reference implementation that covers the most common cases (reading from fs) enabled via a cargo feature.

**Comment by alteous** [comment]
> the user construct Loaded<T> from a specific buffer

glTF data can come from many buffers. For example, a `Primitive` may read from a buffer containing geometry and one containing joints and weights.

> I wouldn't want to be required to load buffers used only for extensions I don't implement

Extensions have always been opt-in via cargo features, so one pays for only what is used. There are no extensions implemented at the moment.

> processing task I'm implementing

Forgive me for being nosy, but what is it you're trying to achieve? glTF is intended to be the 'jpeg of 3D', with the further intention of replacing proprietary asset formats with its own extensible, open format. Hence, supporting conversion from glTF to proprietary formats is not something the crate was ever intended to achieve.

**Comment by Ralith** [comment]
> glTF data can come from many buffers. For example, a Primitive may read from a buffer containing geometry and a separate buffer containing joints and weights.

Ah, that clarifies the reasoning behind the direction you're taking things. I especially like the `Query` trait, though I'd consider returning `vec::IntoIter` instead of `Vec` in preparation for eventually replacing it with `impl Iterator` and removing the need to allocate at all. You could also do something like `fn query<F: FnMut(usize)>(&self, f: F)` but it's probably not worth the decreased ergonomics.

The `Source` trait looks exactly equivalent to `std::ops::Index<usize>` with `type Output=[u8]`. Is there really a need for something distinct?

> Extensions have always been opt-in via cargo features

How are non-standard extensions (which nearly every standard extension starts out as) going to be supported, then? I'm not a big fan of forcing people to fork the library.

> Forgive me for being nosy, but what is it you're trying to achieve?

It was intended as a vague example. A more concrete example might be a tool to extract individual nodes, or a visualizer (or even game engine!) that only loads buffers associated with data that actually needs to be displayed, which, given the "asset library" usecase glTF explicitly supports, should frequently not be "all of them." Even if I *am* using all data in a glTF file, I'd prefer to be able to stream it in continuously (ordered by visibility, say) rather than as a single monolithic blob.

> Hence, supporting conversion from glTF to proprietary formats is not something the crate was ever intended to achieve.

I do, however, fully intend to convert from glTF to a format of my own. Not because I have some weird political requirement to use proprietary formats, but because my internal format has a fundamentally different architecture than glTF does--it's a fine-grained content-addressed store, not to mention that it encodes a variety of application-specific data in addition to geometry and textures.

I'm a fan of glTF's objectives and hope to support it as fully as possible, but it should be no surprise that it's not the perfect terminal format for literally every application.

**Comment by alteous** [comment]
I've abandoned the idea of the `Query` trait for now because one can achieve the same goal without it. Currently, the reference loader just loads everything, guaranteeing all the required data to be available.

**Comment by alteous** [comment]
Updated the OP.

**Comment by alteous** [comment]
The names of some animation data structures are quite verbose. I suggest we rename `InterpolationAlgortihm` as `Interpolation`, `TrsProperty` as `Property` and rename `animation::Target::path` as `animation::Target::property`.

**Comment by alteous** [comment]
https://github.com/KhronosGroup/glTF/pull/1100 suggests we shouldn't be returning `#bin` from `Buffer::uri`.

**Comment by alteous** [comment]
Moving discussion over to #12
